docs/proposals/podpolicy.md

```diff
@@ -0,0 +1,149 @@
+// an empty PodSelector (matches all pods in the namespace).
+type PodPolicySpec struct {
+	// A list of rules that define the policy. Each rule is tested in order, and the first one
+	// that matches is applied. 
+	Rules []PodPolicyRule `json:"rules,omitempty"`
+}
+
+// If the PolicyPredicate is met, then all of the actions specified
+// by the remaining fields are applied. 
+type PodPolicyRule struct {
+	PolicyPredicate *PolicyPredicate `json:"policyPredicate,omitempty"`
+	// Add the specified annotations to the pod.
+	Annotations map[string]string  `json:annotations,omitempty`
+	// In the future, we will add more actions here.
+}
+
+// QoS describes the Quality of Service class of a pod. Kubelet defines QoS on a
+// per-container, per-resource basis, but here we need to match a single QoS per pod.
+// We assume the consumer of PodPolicy will compute the per-pod QoS as the max QoS
+// across all resources and containers in the pod.
+type QoS string
+
+const (
+	QoSGuaranteed QoS = "Guaranteed"
+	QoSBurstable  QoS = "Burstable"
+	QoSBestEffort QoS = "BestEffort"
+)
+
+// All criteria must be met for the PolicyPredicate to be considered met.
+type PolicyPredicate struct {
+	// Note that Namespace is not listed here because PodPolicy is per-Namespace.
+	PodSelector *PodSelector `json:"podSelector,omitempty"`
+	QoS QoS  `json:qos,omitempty`
+}
+
+// PodPolicy describes the PodPolicy API resource, which is used for specifying
+// policies that should be applied to pods based on the pods' characteristics.
+type PodPolicy struct {
+	unversioned.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the pod scheduler policy implemented by this admission controller.
+	// http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+	Spec PodPolicySpec `json:"spec,omitempty"`
+}
+
+// PodPolicyList is a list of PodPolicy items.
+type PodPolicyList struct {
+	unversioned.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#types-kinds
+	unversioned.ListMeta `json:"metadata,omitempty"`
+
+	// Items is a list of PodPolicy objects.
+	// More info: http://releases.k8s.io/HEAD/docs/design/admission_control_resource_quota.md#admissioncontrol-plugin-resourcequota
+	Items []PodPolicy `json:"items"`
+}
+```
+
+## Implementation plan
+
+1. Create `PodPolicy` API resource
+1. Create admission controller that uses policy defined in `PodPolicy` to set scheduler name annotation
+
+*TODO*: What are the semantics of having multiple `PodPolicy` API resources? "Apply all of them" is not
+ideal since there is an ordering to the rules within one. We could do "apply all of them in
+alphabetical order" or something like that...
+
+*TODO*: Is there a way to specify a singleton `PodPolicy` that would apply across all namespaces?
```

Comment:
If it is possible I don't know how.

On Sun, Dec 13, 2015 at 9:27 PM, David Oppenheimer <notifications@github.com
> wrote:

> In docs/proposals/podpolicy.md
> <https://github.com/kubernetes/kubernetes/pull/18262#discussion_r47462485>
> :
>
> > +	// Items is a list of PodPolicy objects.
> > +	// More info: http://releases.k8s.io/HEAD/docs/design/admission_control_resource_quota.md#admissioncontrol-plugin-resourcequota
> > +	Items []PodPolicy `json:"items"`
> > +}
> > +```
> > +
> > +## Implementation plan
> > +
> > +1. Create `PodPolicy` API resource
> > +1. Create admission controller that uses policy defined in `PodPolicy` to set scheduler name annotation
> > +
> > +*TODO*: What are the semantics of having multiple `PodPolicy` API resources? "Apply all of them" is not
> > +ideal since there is an ordering to the rules within one. We could do "apply all of them in
> > +alphabetical order" or something like that...
> > +
> > +*TODO*: Is there a way to specify a singleton `PodPolicy` that would apply across all namespaces?
>
> Should we consider a mechanism for some resources to be namespaced AND
> global? Or maybe some pattern of 2 resources that share the same schema
> but differ only in scope?
>
> Oh, you mean under the current system design, we'd have totally different
> resources (kinds) for global and namespace? I assumed there was some way to
> use a single kind that could be both global and per-namespace and I just
> didn't know how to do it, but I guess you're saying it's not really
> possible right now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18262/files#r47462485>.
>